Authenticate audit events

dividedmind opened the pull request:
This adds 'authenticate' audit events to the audit stream. 

Note the volume of these messages is substantial so we should consider whether they should be enabled by default, or somehow filtered. (These messages weren't present in v4.)

Comments welcome especially on the message schema (see docs/audit.md).

jonahx commented on app/models/audit/event/authn.rb:
```diff
@@ -0,0 +1,53 @@
+require 'ostruct'
+
+class Audit::Event::Authn < OpenStruct
```
Currently the interface for this class is only implicit: you have to read through every method to figure out what attribute assignments must be made before it can be used.  It must be made explicit one way or the other.  My preference would be:
1. Write a normal intializer which accepts the member variables.  This has the advantage of not only being explicit, but also making the client code interface clear and consistent.  There's really no reason anyone should be using attr setters with this object.  Also, you can enforce immutability.
2. If you don't want to do 1. right now,  please at least document the intended interface in comments.

dividedmind replied:
Good point, I'll make it clearer.

jonahx commented on app/models/audit/event/authn.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def service_id
+    service.try :id
```
I'd prefer not to rely on Rails methods in our domain code.  I'd like to see a require active_support if you want to keep it (just to make that dep clear) or else `service.id rescue nil`

dividedmind replied:
I understand where you're coming from, but I like the expressiveness of some of ActiveSupport utilities and this code is unlikely to be moved out.

jonahx commented on app/models/audit/event/authn.rb:
```diff
@@ -0,0 +1,53 @@
+require 'ostruct'
+
+class Audit::Event::Authn < OpenStruct
+  def emit_success
```
You don't need to fix this, but there's something confusing to me about the naming.  we have an "event" emitting a "success" (a "success event").  so an event is emitting an event, which reads oddly.  

There's also a kind of strange inversion in the design, in that this object delegates to the thing that actually writes the log.  Ie, we have an `Event` with an `Audit` (the actually log writer) inside of it, and the event calls its own `emit_success` which then calls Audit, passing the events own structured data to `Audit.info`.  

I think a simpler, clearer design would be to just have pass events to Audit's writer methods, and allow client code to use `Audit`.  But that's a larger (though not very large) refactor -- just food for thought.

dividedmind replied:
Spent some time thinking about it actually. Maybe it'll get refactored at some point.

jonahx commented on app/models/resource.rb:
```diff
@@ -52,6 +52,12 @@ def find_if_visible role, *a
     end
+
+    # Specialization to allow lookup by composite ids,
+    # eg. Resource[account, kind, id]
```
Just a note for later (not specific to your code), but just to make you aware: we have similar logic to this in multiple places in our code base (and others I believe) including my rotator code.  At some point we should DRY it all up.

dividedmind replied:
Agreed. That's why I added this method in lieu of copying the same functionality again, even though I didn't go the one step further and found all the places where it could be used.

jonahx replied:
I don't understand you what you mean, this does copy the same functionality again.  It can't be reused beause here it exists in a very specific context (inside an audit event object), whereas in the codebase in general it's used in many different contexts.  I'm not saying this because I want you to change anything atm, just want to clarify my initial point.

dividedmind replied:
What prevents this method from being used in all the other different contexts?

dividedmind commented on app/models/audit.rb:
```diff
@@ -82,8 +87,10 @@ def call severity, time, progname, msg
       msgid = msg.msgid if msg.respond_to? :msgid
       sd = format_sd sd
 
+      facility = msg.try(:facility) || 4
```
Yeah, I'll probably do that eventually.

dividedmind commented on app/models/audit.rb:
```diff
@@ -82,8 +87,10 @@ def call severity, time, progname, msg
+
       fields = [timestamp, hostname, progname, pid, msgid, sd, msg]
-      ["<#{severity + FACILITY}>1", *fields.map {|x| x || '-'}].join(" ") + "\n"
+      ["<#{severity + (facility << 3)}>1", *fields.map {|x| x || '-'}].join(" ") + "\n"
```
Yeah, this is an RFC thing. I'll change it to `* 8` to make it clearer.

dividedmind commented on app/models/audit/event/authn.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+  end
+
+  SUCCESS_TEMPLATE = "%s successfully authenticated with authenticator %s service %s".freeze
```
If they'd been meant for external use, I would. But since these are in a way only local, I like to keep them near the code that uses them.

dividedmind commented on app/models/audit.rb:
```diff
@@ -31,14 +38,15 @@ def self.conjur_sdid label
   end
 
   class LogMessage < String
```
I'll add it, thanks.

dividedmind commented on app/models/audit.rb:
```diff
@@ -31,14 +38,15 @@ def self.conjur_sdid label
+    attr_reader :msgid, :structured_data, :severity, :facility
   end
 
   # Middleware to store request ID in a thread variable
```
There is no specific reason. It'll be moved out at some point.

kgilpin commented:
Some of your answers to the questions in the PR really seem like they would
be very helpful as comments in the code. For example, why you inherit from
String.
On Tue, Jun 19, 2018 at 5:50 AM Rafał Rzepecki <notifications@github.com>
wrote:

> *@dividedmind* commented on this pull request.
> ------------------------------
>
> In app/models/resource.rb
> <https://github.com/cyberark/conjur/pull/571#discussion_r196360381>:
>
> > @@ -52,6 +52,12 @@ def find_if_visible role, *a
>        res = find *a
>        res if res.try :visible_to?, role
>      end
> +
> +    # Specialization to allow lookup by composite ids,
> +    # eg. Resource[account, kind, id]
>
> Agreed. That's why I added this method in lieu of copying the same
> functionality again, even though I didn't go the one step further and found
> all the places where it could be used.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cyberark/conjur/pull/571#discussion_r196360381>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFRexaFrSfjEtZyG5VsWikfiS5Uw8nEks5t-MlQgaJpZM4UqSSP>
> .
>

dividedmind commented:
Listen @jonahx , I spent all night trying to get to a working solution that would make us both happy, but I spent way too much time on it already and am over any reasonable timebox. Let's leave this as is and please file a tech debt issue if you want.

Not that the time has been wasted, as I added a lot of test code and noticed and fixed a couple bugs on the way, which is why this PR grown a little bit; you might want to give it another look-over. I tried to address some of your concerns (however some are outside the scope of this change).

The commits are clean so you can go through them one-by-one, and also if you're going to merge do either rebase merge or straight merge.

For the record here are WIPs of two approached I tried:
- https://github.com/cyberark/conjur/commit/027b5986d743c2ab3a7a4b5038c76b04372adb24
- https://github.com/cyberark/conjur/commit/f1387b6e6079f4bbae39ec81d68345f69adf0b8f

jvanderhoof commented:
@dividedmind: I like the approach you and Jonah have taken on this one.  I'd ask that you please file issues for work that we reasonably plan to address in the future. As the implementor and visionary of this work, you're in the best place to understand the work that is being pushed out to a future date.

jonahx commented on app/models/audit.rb:
```diff
@@ -21,7 +28,7 @@ module SDID
     CONJUR_PEN = 43868
     def self.conjur_sdid label
-      [label, CONJUR_PEN].join('@').freeze
+      [label, CONJUR_PEN].join('@').intern
```
After your experience with my code, I now believe that any time something in code needs an explanation, having it documented elsewhere is not enough.  Either repeat the explanation directly in the code comments, or include a link to the explanation in the code comments.

dividedmind replied:
Fair :) Could you say more about what do you feel needs explanation?

dividedmind commented on security-scan.sh:
```diff
@@ -18,7 +18,7 @@ done
   mkdir -p brakeman/reports
   chmod -R 0777 brakeman/reports
-  docker run -v "$(pwd):/tmp/" --rm -w /tmp/ codeclimate/codeclimate-brakeman brakeman -o brakeman/reports/brakeman-output.html
+  docker run -v "$(pwd):/tmp/" --rm -w /tmp/ codeclimate/codeclimate-brakeman:b804 brakeman -o brakeman/reports/brakeman-output.html
```
Thank you! I was just about to add this :)

jvanderhoof replied:
No worries. I think we need to start building our own utility containers for stuff like this....

dividedmind replied:
I don't think it's worth the effort for one-tool containers like that. I do think, however, that we ought to start pinning known good versions. Even better, actually review the images to make sure there are no backdoors or surprises of some sort, and then pin by specific SHA that we reviewed.

dividedmind commented:
I need to take a look at why are those new cukes failing in Jenkins. Stay tuned.

dividedmind commented:
(I'll go ahead and merge after I fix that, unless I hear a veto.)
